pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go

```diff
@@ -203,6 +205,49 @@ func (o *followerReadOracle) useClosestOracle(
 // whether a query can be used with a follower read.
 var followerReadOraclePolicy = replicaoracle.RegisterPolicy(newFollowerReadOracle)
 
+type bulkOracle struct {
+	cfg       replicaoracle.Config
+	locFilter roachpb.Locality
+}
+
+var _ replicaoracle.Oracle = bulkOracle{}
+
+// NewBulkOracle returns an oracle for planning bulk operations, which will plan
+// balancing random across all replicas if follower reads are enabled.
+func NewBulkOracle(cfg replicaoracle.Config, locFilter roachpb.Locality) replicaoracle.Oracle {
+	return bulkOracle{cfg: cfg, locFilter: locFilter}
+}
+
+// ChoosePreferredReplica implements the replicaoracle.Oracle interface.
+func (r bulkOracle) ChoosePreferredReplica(
+	ctx context.Context,
+	_ *kv.Txn,
+	desc *roachpb.RangeDescriptor,
+	leaseholder *roachpb.ReplicaDescriptor,
+	_ roachpb.RangeClosedTimestampPolicy,
+	_ replicaoracle.QueryState,
+) (_ roachpb.ReplicaDescriptor, ignoreMisplannedRanges bool, _ error) {
+	if leaseholder != nil && !checkFollowerReadsEnabled(uuid.UUID{} /*not used*/, r.cfg.Settings) {
+		return *leaseholder, false, nil
+	}
+	replicas, err := kvcoord.NewReplicaSlice(ctx, r.cfg.NodeDescs, desc, nil, kvcoord.AllExtantReplicas)
+	if err != nil {
+		return roachpb.ReplicaDescriptor{}, false, sqlerrors.NewRangeUnavailableError(desc.RangeID, err)
+	}
+	if r.locFilter.NonEmpty() {
+		var matches []roachpb.ReplicaDescriptor
+		for i := range replicas {
+			if ok, _ := replicas[i].Locality.Matches(r.locFilter); ok {
+				matches = append(matches, replicas[i].ReplicaDescriptor)
+			}
+		}
+		if len(matches) > 0 {
```

Comment:
I think that's a distsql level decision: it can and already does look at the node that this function picks for one range and then see if can or can't use that node, and keeps stats about what nodes are getting used and why. I don't really think it is worth re-doing that logic down here.